feat: batched and chunked insertion and deletion of relation tuples and UUID mappings

alnr opened the pull request.

aeneasr reviewed (changes requested):
Generally looks good but I suggest we add tests that delete 500 tuples and create 500 tuples because my gut feeling says that we'll run into issues.

alnr commented on internal/persistence/sql/relationtuples.go:
```diff
@@ -165,25 +142,53 @@ func (p *Persister) whereQuery(ctx context.Context, q *pop.Query, rq *relationtu
+			ors = append(ors, "(namespace = ? AND object = ? AND relation = ? AND subject_id = ? AND subject_set_namespace IS NULL AND subject_set_object IS NULL AND subject_set_relation IS NULL)")
+			args = append(args, rt.Namespace, rt.Object, rt.Relation, s.ID)
+		case *relationtuple.SubjectSet:
+			ors = append(ors, "(namespace = ? AND object = ? AND relation = ? AND subject_id IS NULL AND subject_set_namespace = ? AND subject_set_object = ? AND subject_set_relation = ?)")
```
Are you suggesting to use the pop query builder? Not possible because it can only `AND` several `WHERE` clauses together. There is no `OR`.

aeneasr replied:
Sorry no, this was just a note to myself to comapre the queries!

alnr commented on internal/persistence/sql/relationtuples.go:
```diff
@@ -165,25 +142,53 @@ func (p *Persister) whereQuery(ctx context.Context, q *pop.Query, rq *relationtu
-				Where("object = ?", r.Object).
-				Where("relation = ?", r.Relation)
-			if err := p.whereSubject(ctx, q, r.Subject); err != nil {
+		for chunk := range slices.Chunk(rs, 500) {
```
100 OK?

zepatrik replied:
I think the more you can use in one query the better, because we'll save round-trips. Just add a test that reaches the maximum and see if it passes?

aeneasr replied:
Yeah agree!

alnr replied:
Funnily, we don't seem to have an explicit limit on how many tuples we allow you to insert in one API call. gRPC has a max request size, but HTTP doesn't of course.

I've added a test which led me to discover that inserting the UUID mappings was batched (one statement) but not chunked. So I fixed that too.

Added comments and constants explaining chunk sizing.

IMO we should limit the max size also in the handlers. But the DB layer is capable of a lot!

aeneasr commented on internal/persistence/sql/query_test.go:
```diff
@@ -0,0 +1,49 @@
+	require.NoError(t, err)
+
+	// parentheses are important here
+	assert.Equal(t, q, "DELETE FROM keto_relation_tuples WHERE ((namespace = ? AND object = ? AND relation = ? AND subject_id = ? AND subject_set_namespace IS NULL AND subject_set_object IS NULL AND subject_set_relation IS NULL) OR (namespace = ? AND object = ? AND relation = ? AND subject_id IS NULL AND subject_set_namespace = ? AND subject_set_object = ? AND subject_set_relation = ?)) AND nid = ?" ...
```
I spoke to @zepatrik and he said there is no other way to do this, unfortunately. So I guess here too we just test what the maximum amount of parameters is and then set the limit accordingly?

alnr commented on internal/persistence/sql/relationtuples.go:
```diff
@@ -260,15 +265,63 @@ func (p *Persister) ExistsRelationTuples(ctx context.Context, query *relationtup
 	return exists, sqlcon.HandleError(err)
 }
 
+func buildInsert(commitTime time.Time, nid uuid.UUID, rs []*relationtuple.RelationTuple) (query string, args []any, err error) {
```
done

zepatrik commented on internal/persistence/sql/relationtuples.go:
```diff
@@ -6,18 +6,29 @@ package sql
+const (
+	chunkSizeInsertUUIDMappings = 15000 // two placeholders per mapping
+	chunkSizeInsertTuple        = 3000  // ten placeholders per tuple
+	chunkSizeDeleteTuple        = 100   // the database must build an expression tree for each chunk, so we must limit more agressively
```
What exactly do you mean by "build an expression tree" here?

alnr replied:
The DB parses the `DELETE WHERE` query and will construct a data structure for it. That's likely not an inexpensive operation especially if optimizations are performed on that structure afterwards. SQLite actually has a limit of 1000 nodes for this structure.

Inserts are a lot simpler in this regard, and we can have bigger queries.

zepatrik commented on internal/persistence/sql/relationtuples.go:
```diff
@@ -260,15 +272,63 @@ func (p *Persister) ExistsRelationTuples(ctx context.Context, query *relationtup
+		}
+
+		rt := &RelationTuple{
+			ID:         uuid.Must(uuid.NewV4()),
```
Use `gen_random_uuid()` on crdb?

alnr replied:
We could, but performance is not likely to increase a lot because these are global tables anyway.

zepatrik commented on internal/persistence/sql/relationtuples.go:
```diff
@@ -260,15 +272,63 @@ func (p *Persister) ExistsRelationTuples(ctx context.Context, query *relationtup
+	}
+
+	var q strings.Builder
+	fmt.Fprintf(&q, "INSERT INTO %s (shard_id, nid, namespace, object, relation, subject_id, subject_set_namespace, subject_set_object, subject_set_relation, commit_time) VALUES ", (&RelationTuple{}).TableName())
```
As the nid and commit time are always the same, we could do this optimized query instead (simplified example):
```sql
INSERT INTO vals (nid, id, s)
SELECT 
    $1 AS nid, 
    gen_random_uuid() AS id, 
    v.s
FROM 
    (VALUES ($2), ($3), ($3)) AS v(s);
```
Not a blocker IMO, I'm happy either way.

alnr replied:
I don't think it is valid SQL to have `SELECT $1`. Either way, the chunk size is IMO large enough.

zepatrik commented on internal/e2e/transaction_cases_test.go:
```diff
@@ -58,6 +60,53 @@ func runTransactionCases(c transactClient, m *namespaceTestManager) func(*testin
+			resp := c.queryTuple(t, &ketoapi.RelationQuery{
+				Namespace: &ns[0].Name,
+			})
+			// Use this instead, but currently unbearably huge output
+			// assert.ElementsMatch(t, tuples, resp.RelationTuples)
+			for i := range tuples {
+				require.Contains(t, resp.RelationTuples, tuples[i])
+			}
```
Sorry to break your test here, but it was not properly testing the functionality, as all RTs were the same.
It looks like you have to add pagination here to get the full list of relation tuples now (which you should have done from the beginning).

alnr replied:
pagination added!

zepatrik reviewed (approved):
Great 🎉

alnr commented:
Alright after being led down some rabbit holes following red herrings (tired of animals yet?) I fixed the test and then marked it as slow.

But while investigating I also cleaned up some deprecated and arguably incorrect uses of the Go gRPC API and fixed a race condition. Didn't touch the non-test code.

Should be good to merge.

AustinCase commented:
Hey! @aeneasr @alnr and @zepatrik saw this merged, but it looks like the last release thats available is v0.13-alpha from feb of last year. Are those being kept up to date?

aeneasr commented:
We definitely owe a release, Patrik and me are both PTO but I‘ll put it on my list for when I‘m back end of Feb
